components/raftstore/src/store/util.rs

```diff
@@ -836,6 +837,155 @@ impl Display for MsgType<'_> {
     }
 }
 
+/// `RegionReadProgress` is used to keep track of the replica's `safe_ts`, the replica can handle a read
+/// request directly without requiring leader lease or read index iff `safe_ts` >= `read_ts` (the `read_ts`
+/// is usually stale i.e seconds ago).
+///
+/// `safe_ts` is updated by the `(apply index, safe ts)` item:
+/// ```
+/// if self.applied_index >= item.apply_index {
+///     self.safe_ts = max(self.safe_ts, item.safe_ts)
+/// }
+/// ```
+///
+/// For the leader, the `(apply index, safe ts)` item is publish by the `resolved-ts` worker periodically.
+/// For the followers, the item is sync periodically from the leader through the `CheckLeader` rpc.
+///
+/// The intend is to make the item's `safe ts` larger (more up to date) and `apply index` smaller (require less data)
+#[derive(Debug)]
+pub struct RegionReadProgress {
+    // `core` used to keep track and update `safe_ts`, it should
+    // not be accessed outside to avoid dead lock
+    core: Mutex<RegionReadProgressCore>,
+    // The fast path to read `safe_ts` and `applied_index` without
+    // acquiring the mutex on `core`
+    safe_ts: AtomicU64,
+    applied_index: AtomicU64,
+}
+
+impl RegionReadProgress {
+    pub fn new(applied_index: u64, cap: usize) -> RegionReadProgress {
+        RegionReadProgress {
+            core: Mutex::new(RegionReadProgressCore::new(applied_index, cap)),
+            applied_index: AtomicU64::from(applied_index),
+            safe_ts: AtomicU64::from(0),
+        }
+    }
+
+    pub fn update_applied(&self, applied: u64) {
+        let mut core = self.core.lock().unwrap();
+        if let Some(ts) = core.update_applied(applied) {
+            self.safe_ts.fetch_max(ts, AtomicOrdering::SeqCst);
```

Comment:
Is it possible to go backward?

Reply:
It won't in this place, using `fetch_max` here for robustness.

Reply:
Just `store` should be enough given that `update_applied` won't give a smaller ts.